docs/architecture/adr-038-state-listening.md

```diff
@@ -0,0 +1,584 @@
+	writer io.Writer
+	prefixBuf [6]byte
+}
+
+// NewPrefixWriteListener wraps a PrefixWriteListener around an io.Writer
+func NewPrefixWriteListener(w io.Writer) *PrefixWriteListener {
+	return &PrefixWriteListener{
+		writer: w,
+	}
+}
+
+// OnWrite satisfies the WriteListener interface by writing out big endian length-prefixed key-value pairs
+// to an underlying io.Writer
+// The first two bytes of the prefix encode the length of the key
+// The last four bytes of the prefix encode the length of the value
+// This WriteListener makes two assumptions
+// 1) The key is no longer than 1<<16 - 1
+// 2) The value is no longer than 1<<32 - 1
+func (swl *PrefixWriteListener) OnWrite(key []byte, value []byte) {
+	keyLen := len(key)
+	valLen := len(key)
+	if keyLen > math.MaxUint16 || valLen > math.MaxUint32 {
+		return
+	}
+	binary.BigEndian.PutUint16(l.prefixBuf[:2], uint16(keyLen))
+	binary.BigEndian.PutUint32(l.prefixBuf[2:], uint32(valLen))
+	l.writer.Write(l.prefixBuf[:])
+	l.writer.Write(key)
+	l.writer.Write(value)
+}
+```
+
+And one that writes out newline-delineated key-length-prefixed key-value pairs to an underlying io.Writer:
+
+```go
+// NewlineWriteListener is used to configure listening to a KVStore by writing out big endian key-length-prefixed and
+// newline delineated key-value pairs to an io.Writer
+type NewlineWriteListener struct {
+	writer              io.Writer
+	keyLenBuf           [2]byte
+}
+
+// NewNewlineWriteListener wraps a StockWriteListener around an io.Writer
+func NewNewlineWriteListener(w io.Writer) *NewlineWriteListener {
+	return &NewlineWriteListener{
+		writer: w,
+	}
+}
+
+var newline = []byte("\n")
+
+// OnWrite satisfies WriteListener interface by writing out newline delineated big endian key-length-prefixed key-value
+// pairs to the underlying io.Writer
+// The first two bytes encode the length of the key
+// Separate key-value pairs are newline delineated
+// This WriteListener makes three assumptions
+// 1) The key is no longer than 1<<16 - 1
+// 2) The value and keys contain no newline characters
+func (l *NewlineWriteListener) OnWrite(key []byte, value []byte) {
+	keyLen := len(key)
+	if keyLen > math.MaxUint16 {
+		return
+	}
+	binary.BigEndian.PutUint16(l.keyLenBuf[:], uint16(keyLen))
+	l.writer.Write(e.keyLenBuf[:])
+	l.writer.Write(key)
+	l.writer.Write(value)
+	l.writer.Write(newline)
+}
+```
+
+The former makes no assumptions about the presence of newline characters in keys or values, but values
+must be no longer than 1<<32 - 1. The latter assumes newlines are not present in keys or values but can support any length
+of value. Both assume keys are no longer than 1<<16 - 1. Newline delineation improves readability by enabling a consumer to orient
+themselves at the start of a key-value pair at any point in the stream (e.g. tail a file), without character delineation a consumer must start
+at the beginning of the stream and not lose track of their position in the stream.
```

Comment:
As a counter argument, I would say this adds additional work. If there are `N` WriteListeners and `M` StreamingServices, that's potentially `N*M` connectors/adaptors to maintain.

What do you think of using this ADR to define one single spec for what the WriteListener writes, and consumers can get creative with StreamingServices by reading bytes from this spec.

Reply:
I like that idea, I feel like the scope creep is catching up on this ADR (my bad). 

As a compromise this ADR can be focused on the `WriteListener` and the single file streaming service that @aaronc outlines above. Using that approach to writing to files, we can abandon the need for newline delineation as instead delineation occurs at the file level, e.g. it's okay to have to always have to start at the head of the file if each file is only for a single ABCI stage of a single block.